src/NLog/Targets/FileArchiveStyleFactory.cs

```diff
@@ -0,0 +1,758 @@
+        /// </summary>
+        public const string PatternEndCharacters = "#}";
+
+        /// <summary>
+        /// File name which is used as template for matching and replacements. 
+        /// It is expected to contain a pattern to match.
+        /// </summary>
+        public string Template { get { return this.template; } }
+
+        /// <summary>
+        /// The begging position of the <see cref="P:FileNameTemplate.Pattern"/> 
+        /// within the <see cref="P:FileNameTemplate.Template"/>. -1 is returned 
+        /// when no pattern can be found.
+        /// </summary>
+        public int BeginAt { get { return startIndex; } }
+
+        /// <summary>
+        /// The ending position of the <see cref="P:FileNameTemplate.Pattern"/> 
+        /// within the <see cref="P:FileNameTemplate.Template"/>. -1 is returned 
+        /// when no pattern can be found.
+        /// </summary>
+        public int EndAt { get { return endIndex; } }
+
+        private bool FoundPattern { get { return startIndex != -1 && endIndex != -1; } }
+
+        private readonly string template;
+        private readonly int startIndex;
+        private readonly int endIndex;
+
+        public FileNameTemplate(string template)
+        {
+            this.template = template;
+            this.startIndex = template.IndexOf(PatternStartCharacters, StringComparison.Ordinal);
+            if (this.startIndex != -1)
+                this.endIndex = template.IndexOf(PatternEndCharacters, StringComparison.Ordinal) + PatternEndCharacters.Length;
+        }
+
+        /// <summary>
+        /// Replace the pattern with the specified String.
+        /// </summary>
+        /// <param name="replacementValue"></param>
+        /// <returns></returns>
+        public string ReplacePattern(string replacementValue)
+        {
+            return !FoundPattern || String.IsNullOrEmpty(replacementValue) ? this.Template : template.Substring(0, this.BeginAt) + replacementValue + template.Substring(this.EndAt);
+        }
+    }
+
+    abstract class FileArchiveModeBase : IFileArchiveStyle
+    {
+        private string _lastArchiveFilePath;
+
+        /// <summary>
+        /// Check if cleanup should be performed on initialize new file
+        /// 
+        /// Skip cleanup when initializing new file, just after having performed archive operation
+        /// </summary>
+        /// <param name="archiveFilePath">Base archive file pattern</param>
+        /// <returns>True, when archive cleanup is needed</returns>
+        public bool AttemptCleanupOnInitializeFile(string archiveFilePath)
+        {
+            if (_lastArchiveFilePath == archiveFilePath)
+            {
+                // Skip archive cleanup, when it has just been executed
+                _lastArchiveFilePath = string.Empty;
+                return false;
+            }
+            else
+            {
+                _lastArchiveFilePath = string.Empty;
```

Comment:
please move before the `if` (as both branches have ` _lastArchiveFilePath = string.Empty;` - or is this a  potential bug?)